Automattic/sensei
# Add Tools page and deprecate Data Updates

### Changes proposed in this Pull Request

* Introduces a Tools page to replace `Data Updates`.
* Deprecated all data updates. 
  * I went through each of them and found most were destructive and/or unnecessary. 
  * Since we weren't maintaining them, I found several issues that would have caused bugs if ran. 
  * Some might have some usefulness for bulk one-time actions (setting all questions to be worth 1 point), but probably limited and again we haven't been maintaining them. I removed them because they seemed like unnecessary technical debt. Many could be turned into snippets if we think that would be useful.
  * For users who need to upgrade from very old versions (pre-1.9), these would have had to get ran manually anyway. If users _do_ need to upgrade from a pre-v1.9 version, they can upgrade to 3.6 and then run the necessary data updates manually (no guarantee these will work, though).
* I created two tools from deprecated data updates: one to set up roles and capabilities again if somehow they got messed up and another to delete progress data from deleted users. Since v3.0 we've done this automatically, but figured some folks might still have a need for this.
* Moves the main role/capability setting methods to `Sensei_Main`. For these methods, I kept the deprecated functionality (pointed to their new locations) because they were deprecated methods used more directly within Sensei outside of data updates.

### Other Notes
* One todo post merge that I also think we'll want to do to help with the double-progress is **automatically add the indexes to the comments table**. This was a data update tool, but since we stopped doing this automatically sometime pre-1.9, I think this is a big source of the slowdown on lots of Sensei sites. I'll work on this after this PR is merged.
* I think there could be new purpose for the `Sensei_Updates` class. I think (in a new PR) we should take care of the indexes and other things on plugin update (per its original intended use).
* Additionally, I'll create an issue to migrate Sensei Certificates custom data updates to be Sensei Tools.
* From my feature plugin, I'll also migrate over the enrollment debug tool, but that will be its own PR.

### Testing instructions

* Generally test the tools to ensure they do what they're supposed to.
* _Remove Deleted User Data_ is the only new tool that has new code. I refactored it so it wouldn't need pagination. It is difficult to test naturally with Sensei 3.x+, but you can manually remove users from the user table and run it to ensure comments data is removed.


### New/Updated Hooks

* `sensei_tools` : Filter on the tools available in Sensei.


### Deprecated Code

All of `Sensei_Updates`, including:
- `Sensei_Updates::sensei_updates_page`
- `Sensei_Updates::function_in_whitelist`
- `Sensei_Updates::update`
- `Sensei_Updates::set_default_quiz_grade_type`
- `Sensei_Updates::set_default_question_type`
- `Sensei_Updates::update_question_answer_data`
- `Sensei_Updates::update_question_grade_points`
- `Sensei_Updates::convert_essay_paste_questions`
- `Sensei_Updates::set_random_question_order`
- `Sensei_Updates::set_default_show_question_count`
- `Sensei_Updates::remove_deleted_user_activity`
- `Sensei_Updates::add_teacher_role`
- `Sensei_Updates::restructure_question_meta`
- `Sensei_Updates::update_quiz_settings`
- `Sensei_Updates::reset_lesson_order_meta`
- `Sensei_Updates::update_question_gap_fill_separators`
- `Sensei_Updates::update_quiz_lesson_relationship`
- `Sensei_Updates::status_changes_fix_lessons`
- `Sensei_Updates::status_changes_convert_lessons`
- `Sensei_Updates::status_changes_convert_courses`
- `Sensei_Updates::status_changes_repair_course_statuses`
- `Sensei_Updates::status_changes_convert_questions`
- `Sensei_Updates::update_legacy_sensei_comments_status`
- `Sensei_Updates::update_comment_course_lesson_comment_counts`
- `Sensei_Updates::remove_legacy_comments`
- `Sensei_Updates::index_comment_status_field`
- `Sensei_Updates::enhance_teacher_role`
- `Sensei_Updates::recalculate_enrolment`

### Screenshot
<img width="1619" alt="Screen Shot 2020-12-10 at 10 37 32 pm" src="https://user-images.githubusercontent.com/68693/101838311-4cf0f800-3b38-11eb-9413-066060f3ff31.png">
